### ISSUE TITLE
[flang][OpenMP] Assertion failure on DECLARE TARGET naming a namelist group or a derived type

### ISSUE THREAD

**Issue by eugeneepshteyn** [llvm/llvm-project#213563]
(Issue created with the help of AI while investigating another flang failure.)

### Summary

`!$omp declare target(...)` crashes flang with an assertion failure when the item named
is a **namelist group** or a **derived type**. Both are invalid `DECLARE TARGET` items, so
flang should emit a semantic diagnostic; instead it aborts.

### Reproducers

Namelist:
```fortran
! namelist group
program p
  real :: x, y
  namelist /n/ x, y
  !$omp declare target(n)
end program
```

Derived type:
```fortran
! derived type name
program p
  type :: t
    integer :: i
  end type
  !$omp declare target(t)
end program
```

```console
$ flang -fopenmp -c repro.f90
```

### Actual behaviour

```
flang: llvm-project/flang/lib/Semantics/resolve-directives.cpp:2407:
  auto Fortran::semantics::OmpAttributeVisitor::Pre(const parser::OmpDeclareTargetDirective &)
    ::(anonymous class)::operator()(auto &) const [d:auto = Fortran::semantics::NamelistDetails]:
  Assertion `(std::is_base_of_v<WithOmpDeclarative, TypeD>) && "Unexpected details type"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ ...
Stack dump:
0.  Program arguments: flang -fc1 -triple x86_64-unknown-linux-gnu -emit-obj -fopenmp ... -x f95 repro.f90
 #11 Fortran::semantics::OmpAttributeVisitor::Pre(Fortran::parser::OmpDeclareTargetDirective const&)
 #12 ParseTreeVisitorLookupScope::Walk<OmpDeclareTargetDirective, OmpAttributeVisitor>(...)
 ...
 #18 Fortran::semantics::ResolveOmpParts(Fortran::semantics::SemanticsContext&,
                                         Fortran::parser::ProgramUnit const&)
 #19 Fortran::semantics::ResolveNamesVisitor::Pre(Fortran::parser::ProgramUnit const&)
```

The derived-type case is the same assertion with `[d:auto = Fortran::semantics::DerivedTypeDetails]`.

### Expected behaviour

A semantic error, as flang already produces for other invalid `DECLARE TARGET` items —
e.g. a module name or a statement function are both diagnosed cleanly today.

### Scope

Reproduces with every clause spelling and every OpenMP version:

| variant | result |
|---|---|
| `declare target(n)` | assertion failure |
| `declare target to(n)` | assertion failure |
| `declare target link(n)` | assertion failure |
| `declare target enter(n)` | assertion failure |
| `-fopenmp-version=45 / 50 / 51 / 52 / 60` | assertion failure in all |

Which symbol kinds reach the crash:

| `DECLARE TARGET` item | behaviour |
|---|---|
| namelist group | **assertion failure** (`NamelistDetails`) |
| derived type name | **assertion failure** (`DerivedTypeDetails`) |
| module name | diagnosed correctly |
| statement function | diagnosed correctly |
| variable / named constant / common block | accepted |

### Regression range

Introduced by 270a6a8f6742:

```
270a6a8f6742 [flang][OpenMP] Store DECLARE_TARGET information in WithOmpDeclarative (#201103)
```

`git blame` attributes the `assert` and its enclosing `else` to that commit.

### Notes

- Only assertions-enabled builds abort.
- Verified against two independently built compilers from different upstream `main`
  bases, both of which reproduce identically.

**Comment by eugeneepshteyn** [comment]
AI claims below have not been verified, but may be useful for issue investigation:

### Root cause

In `OmpAttributeVisitor::Pre(const parser::OmpDeclareTargetDirective &)`
(`flang/lib/Semantics/resolve-directives.cpp`, around lines 2380–2410) the visitor over
`Symbol::details()` has three branches:

```cpp
common::visit(
    [&](auto &d) {
      using TypeD = llvm::remove_cvref_t<decltype(d)>;
      if constexpr (std::is_base_of_v<WithOmpDeclarative, TypeD>) {
        // ... record the DECLARE TARGET information
      } else if constexpr (std::is_same_v<GenericDetails, TypeD> ||
          std::is_same_v<TypeParamDetails, TypeD> ||
          std::is_same_v<MiscDetails, TypeD>) { // e.g. KindParamInquiry
        // These cannot be specified on DECLARE_TARGET. The symbol will
        // receive the OmpDeclareTarget flag (for a diagnostic message),
        // but no details will be modified.
      } else {
        // Catch any unexpected cases.
        assert((std::is_base_of_v<WithOmpDeclarative, TypeD>) &&
            "Unexpected details type");
      }
    },
    const_cast<Symbol *>(sym)->details());
```

The second branch is the intended handling for symbols that cannot legally appear on
`DECLARE TARGET`: leave the details alone and let the later diagnostic fire. But its list
is enumerated explicitly, and `NamelistDetails` and `DerivedTypeDetails` are not in it, so
they fall through to the catch-all `assert`.

Since any name the user writes in the clause is resolved before this point, that
catch-all is reachable from ordinary invalid input rather than only from internal
inconsistency — an assertion is not the right tool here.

### Suggested fix

Either add the missing kinds to the "cannot be specified" branch:

```cpp
      } else if constexpr (std::is_same_v<GenericDetails, TypeD> ||
          std::is_same_v<TypeParamDetails, TypeD> ||
          std::is_same_v<NamelistDetails, TypeD> ||
          std::is_same_v<DerivedTypeDetails, TypeD> ||
          std::is_same_v<MiscDetails, TypeD>) { // e.g. KindParamInquiry
```

or — more robustly, since the branch is user-reachable — drop the `assert` so that any
unhandled details kind degrades to "leave details unmodified and let semantics diagnose"
instead of aborting the compiler.

A dedicated message naming the offending kind (e.g. "DECLARE TARGET item cannot be a
NAMELIST group") would be the friendliest outcome.
